### ISSUE TITLE
When body function use "arguments" object, should there be an error (using "noUnusedParameter" is true) 

### ISSUE THREAD

**Issue by yuit** [Microsoft/TypeScript#9792]
TypeScript : 2.0.0

While trying to apply "noUnusedParameter" option in TypeScript code base, I ran into situation when one use Function.apply and pass in "arguments" object. Our compiler will then issue an error that the containing function's parameter is unused, should we still give an error in this case or by seeing "arguments" object use in function body we should not give an error

**Comment by RyanCavanaugh** [comment]
I would argue this is a bit unusual and the author should write something like this to make their intent clear:
```ts
function fn(arg1, arg2): void;
function fn() {
  // do something with 'arguments' here
}
```

**Comment by RyanCavanaugh** [comment]
... though I guess this doesn't handle the case where you want to use `arg2` but not `arg1`

**Comment by yortus** [comment]
You've got to declare formal parameters if you want the function to have a specific arity (eg `(function fn(a, b, c) {}).length === 3`). But for whatever reason you might process the actual parameters using `arguments`.

Doesn't #9458 solve this? i.e., declare it like `function fn(_arg1, _arg2) {...}`

**Comment by yortus** [comment]
Also, I don't think the mere presence of `arguments` in a function body is enough to satisfy the 'unused parameter' check, since it can appear in arbitrary expressions that in some cases access all the parameters, and in other cases only some of them, or even none of them. You couldn't in general tell one case from the other at compile time.

**Comment by poke** [comment]
Shouldn’t [rest parameters](https://developer.mozilla.org/en/docs/Web/JavaScript/Reference/Functions/rest_parameters) eliminate the need for `arguments` here?

**Comment by mhegazy** [comment]
> Shouldn’t rest parameters eliminate the need for arguments here?

rest parameters will add a for loop and a copy operation to your output. if you are using `arguments`, and this is a performance critical path, i would not use rest args.

**Comment by mhegazy** [comment]
As noted by @RyanCavanaugh and @yortus, you have two options, either remove the arguments and use an overload, or rename them to start with `_`.

**Comment by yortus** [comment]
> you have two options, either remove the arguments and use an overload, or rename them to start with _.

@mhegazy or both! The overload supplies pretty param names for intellisense, and the implementation maintains the correct arity (ie `Function#length`) by declaring formal params. E.g.:

```ts
function fn(arg1, arg2): void; // pretty param names for intellisense
function fn(_1, _2) {  // formal params to preserve function arity
  // do something with 'arguments' here
}```

**Comment by mhegazy** [comment]
Sure, that would work as well.
